src/main/java/org/candlepin/cache/AnonymousCertContentCache.java

```diff
@@ -0,0 +1,137 @@
+
+    @Inject
+    public AnonymousCertContentCache(Configuration config) throws ConfigurationException {
+        Objects.requireNonNull(config);
+
+        long expirationDuration = config.getLong(ConfigProperties.CACHE_ANON_CERT_CONTENT_TTL);
+        if (expirationDuration <= 0) {
+            String msg = ConfigProperties.CACHE_ANON_CERT_CONTENT_TTL + " value must be larger than 0";
+            throw new ConfigurationException(msg);
+        }
+
+        cache = Caffeine.newBuilder()
+            .expireAfterWrite(Duration.ofMillis(expirationDuration))
+            .build();
+    }
+
+    /**
+     * Retrieves cached {@link AnonymousCertContent} for the provided top level SKU IDs.
+     *
+     * @param skuIds
+     *  the top level SKU IDs to retrieve consumer certificate content for
+     *
+     * @throws IllegalArgumentException
+     *  if the provided SKU IDs are null
+     *
+     * @return
+     *  the cached consumer certificate content for the provided top level SKU IDs
+     */
+    public AnonymousCertContent get(Collection<String> skuIds) {
+        if (skuIds == null) {
+            throw new IllegalArgumentException("sku ID is null");
+        }
+
+        if (skuIds.isEmpty()) {
+            return null;
+        }
+
+        return cache.getIfPresent(getKey(skuIds));
+    }
+
+    /**
+     * Inserts {@link AnonymousCertContent} into the cache for the provided top level SKU IDs. An existing
+     * entry in the cache for the same SKU IDs will be replaced.
+     *
+     * @param skuIds
+     *  the top level SKU IDs to associate the consumer certificate content to in the cache
+     *
+     * @param content
+     *  the consumer certificate content to insert into the cache
+     *
+     * @throws IllegalArgumentException
+     *  if the provided SKU IDs are null or the provided consumer certificate content is null
+     */
+    public void put(Collection<String> skuIds, AnonymousCertContent content) {
+        if (skuIds == null) {
+            throw new IllegalArgumentException("sku ID is null");
+        }
+
+        if (content == null) {
+            throw new IllegalArgumentException("content is null");
+        }
+
+        cache.put(getKey(skuIds), content);
+    }
+
+    /**
+     * Removes the cached {@link AnonymousCertContent} associated to the provided top level SKU IDs.
+     *
+     * @param skuIds
+     *  the top level SKU IDs to remove cached consumer certificate content for
+     *
+     * @throws IllegalArgumentException
+     *  if the provided SKU IDs are null
+     */
+    public void remove(Collection<String> skuIds) {
+        if (skuIds == null) {
+            throw new IllegalArgumentException("sku IDs is null");
+        }
+
+        cache.invalidate(getKey(skuIds));
+    }
+
+    /**
+     * Clears all entries in the cache
+     */
+    public void removeAll() {
+        cache.invalidateAll();
+    }
+
+    private int getKey(Collection<String> skuIds) {
```

Comment:
If we are going to with custom keys we should create a class for it to make it explicit. E.G. MultiKey which could define constructors for supported inputs and explicitly define a hashCode.
 
Btw. You could get the same result with new HashSet<>(skuIds).hashCode();

Reply:
This MultiKey class is just a wrapper for a HashSet. What are we getting out of it? Wouldn't an unmodifiableSet be more straightforward as previously suggested?

Reply:
We get an explicit place where the rules for our cache key are defined. We could make the wrapped set unmodifiable, but unless we expose the reference we don't have to.

Reply:
alright